nixos/systemd: conditionally leave out some upstream units

jmbaur opened the pull request:
Some upstream systemd units are conditionally installed into the systemd output, so we must make sure the feature that enables their installation is enabled on our side prior to trying to use them.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jmbaur commented:
gentle ping @NixOS/systemd

ElvishJerricco reviewed (commented):
Do the TPM2 features really require building systemd-boot? There's a lot of TPM2 features in systemd that have nothing to do with systemd-boot. Most of them, in fact.

ElvishJerricco reviewed (changes requested):
Huh, apparently it does. In fact, [it seems to require openssl too](https://github.com/systemd/systemd/blob/d2ce0db7a7f13fdd1d235b4d83d85bf436b39a2e/units/meson.build#L576-L580), which I suppose isn't surprising.

Frankly, I think this means that we should actually have assertions in `systemd/default.nix` like `assert withTpm2Tss -> withBootloader && withOpenSSL;`, rather than adding conditional logic to these modules.

ElvishJerricco commented:
The other, non-tpm2 parts of this PR look good though.

ElvishJerricco commented:
Ah, no, that's wrong. There are parts of systemd that use the TPM2 without requiring boot loader stuff, like systemd-creds. So an assertion in the package expression is wrong.

Regardless, it still seems like we need to add `withOpenSSL` to these checks, right? Maybe we could add an extra passthru like `withTpm2Units = withTpm2Tss && withBootloader && withOpenSSL` or something.

jmbaur committed:
systemd: expose withTpm2Units

Exposes if a special trio of options are enabled in the systemd build,
for convenience when accessing in NixOS modules.

jmbaur committed:
nixos/systemd: conditionally leave out some upstream units

Some upstream systemd units are conditionally installed into the systemd
output, so we must make sure the feature that enables their installation
is enabled on our side prior to trying to use them.

jmbaur commented:
@ElvishJerricco applied your suggestions! Added changes to the systemd packge, though they all exist in passthru, so we shouldn't incur the cost of a rebuild.
